File: Java/src/com/dbc/algorithm/Pointer/LongestRepeatingCharacterReplacement.java
package com.dbc.algorithm.Pointer;

public class LongestRepeatingCharacterReplacement {
    public static int characterReplacement(String s, int k) {
        int[] nums = new int[26];
        int left = 0, right = 0, max = 0;
        while (right < s.length()) {
            nums[s.charAt(right) - 'A'] +=1;
            max = Math.max(max, nums[s.charAt(right) - 'A']);
            if (right - left + 1 - max > k) {
                nums[s.charAt(left) - 'A'] -= 1;
                left++;
            }
            right++;
        }

        return right - left;
    }
}

File: src/tmp/Python/pointer/longest_repeating_character_replacement.py
def characterReplacement(s: str, k: int) -> int:
    num = [0] * 26
    n = len(s)
    maxn = left = right = 0
    while right < n:
        num[ord(s[right]) - ord("A")] += 1
        maxn = max(maxn, num[ord(s[right]) - ord("A")])
        if right - left + 1 - maxn > k:
            num[ord(s[left]) - ord("A")] -= 1
            left += 1
        right += 1
    
    return right - left

if __name__ == "__main__":
    print(characterReplacement(s = "ABAB", k = 2))
        